File: app/src/main/java/com/robifr/ledger/ui/createqueue/CreateQueueProductOrder.java
/**
 * Copyright (c) 2024 Robi
 *
 * Ledger is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Ledger is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Ledger. If not, see <https://www.gnu.org/licenses/>.
 */

package com.robifr.ledger.ui.createqueue;

import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.LinearLayout;
import androidx.annotation.ColorInt;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.view.ActionMode;
import com.google.android.material.card.MaterialCardView;
import com.robifr.ledger.R;
import com.robifr.ledger.data.model.ProductOrderModel;
import com.robifr.ledger.databinding.ProductOrderCardBinding;
import com.robifr.ledger.util.CurrencyFormat;
import java.math.BigDecimal;
import java.util.List;
import java.util.Objects;
import java.util.Set;
import java.util.TreeSet;

public class CreateQueueProductOrder implements View.OnClickListener, View.OnLongClickListener {
  @NonNull private final CreateQueueFragment _fragment;
  @NonNull private final MakeProductOrder _makeProductOrder;
  @Nullable private ActionMode _contextualMode;

  public CreateQueueProductOrder(@NonNull CreateQueueFragment fragment) {
    this._fragment = Objects.requireNonNull(fragment);
    this._makeProductOrder = new MakeProductOrder(this._fragment);

    this._fragment
        .fragmentBinding()
        .productOrder
        .addButton
        .setOnClickListener(button -> this._makeProductOrder.openCreateDialog());
  }

  @Override
  public void onClick(@NonNull View view) {
    Objects.requireNonNull(view);

    switch (view.getId()) {
      case R.id.cardView -> {
        final ProductOrderCardBinding cardBinding = ProductOrderCardBinding.bind(view);
        final int cardIndex =
            this._fragment
                .fragmentBinding()
                .productOrder
                .listLayout
                .indexOfChild(cardBinding.cardView);
        final Boolean isContextualModeActive =
            this._fragment
                .createQueueViewModel()
                .selectProductOrderView()
                .isContextualModeActive()
                .getValue();

        // Check product order on contextual mode.
        if (isContextualModeActive != null && isContextualModeActive) {
          this._fragment
              .createQueueViewModel()
              .selectProductOrderView()
              .onProductOrderCheckedChanged(cardIndex, !cardBinding.checkbox.isChecked());

          // Edit selected product order.
        } else {
          final List<ProductOrderModel> productOrders =
              this._fragment.createQueueViewModel().inputtedProductOrders().getValue();
          if (productOrders == null) return;

          this._fragment
              .createQueueViewModel()
              .makeProductOrderView()
              .onProductOrderToEditChanged(productOrders.get(cardIndex));
          this._makeProductOrder.openEditDialog();
        }
      }
    }
  }

  @Override
  public boolean onLongClick(@NonNull View view) {
    Objects.requireNonNull(view);

    return switch (view.getId()) {
      case R.id.cardView -> {
        final ProductOrderCardBinding cardBinding = ProductOrderCardBinding.bind(view);
        final int cardIndex =
            this._fragment
                .fragmentBinding()
                .productOrder
                .listLayout
                .indexOfChild(cardBinding.cardView);
        this._fragment
            .createQueueViewModel()
            .selectProductOrderView()
            .onProductOrderCheckedChanged(cardIndex, !cardBinding.checkbox.isChecked());
        yield true;
      }

      default -> false;
    };
  }

  @NonNull
  public MakeProductOrder makeProductOrder() {
    return this._makeProductOrder;
  }

  public void setSelectedProductOrderByIndexes(@NonNull Set<Integer> selectedIndexes) {
    Objects.requireNonNull(selectedIndexes);

    if (this._contextualMode != null) {
      this._contextualMode.setTitle(Integer.toString(selectedIndexes.size()));
    }

    final TreeSet<Integer> indexes = new TreeSet<>(selectedIndexes);

    for (int i = this._fragment.fragmentBinding().productOrder.listLayout.getChildCount();
        i-- > 0; ) {
      final ProductOrderCardBinding cardBinding =
          ProductOrderCardBinding.bind(
              this._fragment.fragmentBinding().productOrder.listLayout.getChildAt(i));
      final boolean shouldCheck = !indexes.isEmpty() && indexes.last() == i;
      final int checkboxVisibility = shouldCheck ? View.VISIBLE : View.GONE;
      final int productImageVisibility = shouldCheck ? View.GONE : View.VISIBLE;

      cardBinding.cardView.setChecked(shouldCheck);
      cardBinding.checkbox.setChecked(shouldCheck);
      cardBinding.checkbox.setVisibility(checkboxVisibility);
      cardBinding.productImage.text.setVisibility(productImageVisibility);

      if (shouldCheck) indexes.pollLast();
    }
  }

  public void setContextualMode(boolean isActive) {
    if (isActive && this._fragment.requireActivity() instanceof AppCompatActivity activity) {
      this._contextualMode =
          activity.startSupportActionMode(new ContextualActionMode(this._fragment));

    } else if (!isActive && this._contextualMode != null) {
      this._contextualMode.finish();
      this._contextualMode = null;
    }
  }

  public void setGrandTotalPrice(@NonNull BigDecimal grandTotalPrice) {
    Objects.requireNonNull(grandTotalPrice);

    this._fragment
        .fragmentBinding()
        .productOrder
        .grandTotalPrice
        .setText(CurrencyFormat.format(grandTotalPrice, "id", "ID"));
  }

  public void setTotalDiscount(@NonNull BigDecimal totalDiscount) {
    Objects.requireNonNull(totalDiscount);

    this._fragment
        .fragmentBinding()
        .productOrder
        .totalDiscount
        .setText(CurrencyFormat.format(totalDiscount, "id", "ID"));
  }

  public void setCustomerBalanceAfterPaymentTitle(@Nullable String title) {
    final int viewVisibility = title != null ? View.VISIBLE : View.GONE;

    this._fragment.fragmentBinding().productOrder.customerBalanceTitle.setText(title);
    this._fragment
        .fragmentBinding()
        .productOrder
        .customerBalanceTitle
        .setVisibility(viewVisibility);
  }

  public void setCustomerBalanceAfterPayment(@Nullable Long balance) {
    final String text =
        balance != null ? CurrencyFormat.format(BigDecimal.valueOf(balance), "id", "ID") : null;
    final int viewVisibility = balance != null ? View.VISIBLE : View.GONE;

    this._fragment.fragmentBinding().productOrder.customerBalance.setText(text);
    this._fragment.fragmentBinding().productOrder.customerBalance.setVisibility(viewVisibility);
  }

  public void setCustomerDebtAfterPaymentTitle(@Nullable String title) {
    final int viewVisibility = title != null ? View.VISIBLE : View.GONE;

    this._fragment.fragmentBinding().productOrder.customerDebtTitle.setText(title);
    this._fragment.fragmentBinding().productOrder.customerDebtTitle.setVisibility(viewVisibility);
  }

  public void setCustomerDebtAfterPayment(@Nullable BigDecimal debt) {
    final String text = debt != null ? CurrencyFormat.format(debt, "id", "ID") : null;
    final int textColor =
        debt != null && debt.compareTo(BigDecimal.ZERO) < 0
            // Negative debt will be shown red.
            ? this._fragment.requireContext().getColor(R.color.red)
            : this._fragment.requireContext().getColor(R.color.text_enabled);
    final int viewVisibility = debt != null ? View.VISIBLE : View.GONE;

    this._fragment.fragmentBinding().productOrder.customerDebt.setText(text);
    this._fragment.fragmentBinding().productOrder.customerDebt.setTextColor(textColor);
    this._fragment.fragmentBinding().productOrder.customerDebt.setVisibility(viewVisibility);
  }

  public void setInputtedProductOrders(@NonNull List<ProductOrderModel> productOrders) {
    Objects.requireNonNull(productOrders);

    final LinearLayout listLayout = this._fragment.fragmentBinding().productOrder.listLayout;

    if (listLayout.getChildCount() > productOrders.size()) {
      for (int i = listLayout.getChildCount(); i-- > 0; ) {
        listLayout.removeViewAt(i); // Remove all the extra views.
      }
    }

    for (int i = 0; i < productOrders.size(); i++) {
      final ProductOrderCardBinding cardBinding =
          // Reuse already inflated view if available to reduce overhead.
          listLayout.getChildAt(i) instanceof MaterialCardView cardView
              ? ProductOrderCardBinding.bind(cardView)
              : ProductOrderCardBinding.inflate(
                  this._fragment.getLayoutInflater(),
                  this._fragment.fragmentBinding().productOrder.listLayout,
                  false);
      final ProductOrderCardComponent components =
          new ProductOrderCardComponent(this._fragment.requireContext(), cardBinding);

      components.setProductOrder(productOrders.get(i));
      cardBinding.cardView.setOnClickListener(this);
      cardBinding.cardView.setOnLongClickListener(this);
      cardBinding.checkbox.setChecked(false);
      cardBinding.checkbox.setVisibility(View.GONE);
      cardBinding.productImage.text.setVisibility(View.VISIBLE);

      // Add to layout until the amount of child view match the orders size.
      if (listLayout.getChildCount() < i + 1) listLayout.addView(cardBinding.getRoot());
    }
  }

  private static class ContextualActionMode implements ActionMode.Callback {
    @NonNull private final CreateQueueFragment _fragment;
    @ColorInt private final int _normalStatusBarColor;

    public ContextualActionMode(@NonNull CreateQueueFragment fragment) {
      this._fragment = Objects.requireNonNull(fragment);
      this._normalStatusBarColor = fragment.requireActivity().getWindow().getStatusBarColor();
    }

    @Override
    public boolean onCreateActionMode(@NonNull ActionMode mode, @NonNull Menu menu) {
      Objects.requireNonNull(mode);
      Objects.requireNonNull(menu);

      mode.getMenuInflater().inflate(R.menu.createqueue_contextualtoolbar, menu);
      // Match status bar color with the contextual toolbar background color.
      this._fragment
          .requireActivity()
          .getWindow()
          .setStatusBarColor(this._fragment.requireContext().getColor(R.color.surface));
      return true;
    }

    @Override
    public boolean onPrepareActionMode(@NonNull ActionMode mode, @NonNull Menu menu) {
      return false;
    }

    @Override
    public boolean onActionItemClicked(@NonNull ActionMode mode, @NonNull MenuItem item) {
      Objects.requireNonNull(mode);
      Objects.requireNonNull(item);

      return switch (item.getItemId()) {
        case R.id.createqueue_contextualtoolbar_delete_item -> {
          this._fragment
              .createQueueViewModel()
              .selectProductOrderView()
              .onDeleteSelectedProductOrder();
          yield true;
        }

        default -> false;
      };
    }

    @Override
    public void onDestroyActionMode(@NonNull ActionMode mode) {
      Objects.requireNonNull(mode);

      this._fragment.createQueueViewModel().selectProductOrderView().reset();
      // Re-apply original status bar color.
      this._fragment.requireActivity().getWindow().setStatusBarColor(this._normalStatusBarColor);
    }
  }
}

File: app/src/main/java/com/robifr/ledger/ui/createqueue/CreateQueueProductOrder.kt
/**
 * Copyright 2024 Robi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.robifr.ledger.ui.createqueue

import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.LinearLayout
import androidx.annotation.ColorInt
import androidx.annotation.ColorRes
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.app.AppCompatDelegate
import androidx.appcompat.view.ActionMode
import com.google.android.material.card.MaterialCardView
import com.robifr.ledger.R
import com.robifr.ledger.data.model.ProductOrderModel
import com.robifr.ledger.databinding.ProductOrderCardBinding
import com.robifr.ledger.util.CurrencyFormat
import java.math.BigDecimal
import java.util.TreeSet

class CreateQueueProductOrder(private val _fragment: CreateQueueFragment) {
  val makeProductOrder: CreateQueueMakeProductOrder = CreateQueueMakeProductOrder(_fragment)
  private var _contextualMode: ActionMode? = null
  @ColorInt
  private val _normalStatusBarColor: Int = _fragment.requireActivity().window.statusBarColor

  init {
    _fragment.fragmentBinding.productOrder.addButton.setOnClickListener {
      _fragment.createQueueViewModel.makeProductOrderView.onShowDialog()
    }
  }

  fun setSelectedProductOrderByIndexes(selectedIndexes: Set<Int>) {
    _contextualMode?.setTitle(selectedIndexes.size.toString())

    val indexes: TreeSet<Int> = TreeSet(selectedIndexes)
    for (i in _fragment.fragmentBinding.productOrder.listLayout.childCount - 1 downTo 0) {
      val shouldCheck: Boolean = indexes.isNotEmpty() && indexes.last() == i
      with(
          ProductOrderCardBinding.bind(
              _fragment.fragmentBinding.productOrder.listLayout.getChildAt(i))) {
            cardView.setChecked(shouldCheck)
            checkbox.setChecked(shouldCheck)
            checkbox.setVisibility(if (shouldCheck) View.VISIBLE else View.GONE)
            productImage.text.setVisibility(if (shouldCheck) View.GONE else View.VISIBLE)
          }
      if (shouldCheck) indexes.pollLast()
    }
  }

  fun setContextualMode(isActive: Boolean) {
    // The null check on contextual mode meant to prevent them from restarting.
    if (isActive && _fragment.requireActivity() is AppCompatActivity && _contextualMode == null) {
      // Match status bar color with the contextual toolbar background color.
      _fragment
          .requireActivity()
          .window
          .setStatusBarColor(_fragment.requireContext().getColor(R.color.surface))
      _contextualMode =
          (_fragment.requireActivity() as AppCompatActivity).startSupportActionMode(
              SelectProductOrderActionModeCallback(_fragment))
    } else if (!isActive && _contextualMode != null) {
      // Re-apply original status bar color.
      _fragment.requireActivity().window.setStatusBarColor(_normalStatusBarColor)
      _contextualMode?.finish()
      _contextualMode = null
    }
  }

  fun setGrandTotalPrice(grandTotalPrice: BigDecimal) {
    _fragment.fragmentBinding.productOrder.grandTotalPrice.setText(
        CurrencyFormat.format(
            grandTotalPrice, AppCompatDelegate.getApplicationLocales().toLanguageTags()))
  }

  fun setTotalDiscount(totalDiscount: BigDecimal) {
    _fragment.fragmentBinding.productOrder.totalDiscount.setText(
        CurrencyFormat.format(
            totalDiscount, AppCompatDelegate.getApplicationLocales().toLanguageTags()))
  }

  fun setCustomerBalanceAfterPaymentTitle(customerName: String?, isVisible: Boolean) {
    _fragment.fragmentBinding.productOrder.customerBalanceTitle.setText(
        _fragment.getString(R.string.createQueue_productOrders_x_balance, customerName))
    _fragment.fragmentBinding.productOrder.customerBalanceTitle.setVisibility(
        if (isVisible) View.VISIBLE else View.GONE)
  }

  fun setCustomerBalanceAfterPayment(balance: Long?, isVisible: Boolean) {
    _fragment.fragmentBinding.productOrder.customerBalance.setText(
        balance?.let {
          CurrencyFormat.format(
              it.toBigDecimal(), AppCompatDelegate.getApplicationLocales().toLanguageTags())
        })
    _fragment.fragmentBinding.productOrder.customerBalance.setVisibility(
        if (isVisible) View.VISIBLE else View.GONE)
  }

  fun setCustomerDebtAfterPaymentTitle(customerName: String?, isVisible: Boolean) {
    _fragment.fragmentBinding.productOrder.customerDebtTitle.setText(
        _fragment.getString(R.string.createQueue_productOrders_x_debt, customerName))
    _fragment.fragmentBinding.productOrder.customerDebtTitle.setVisibility(
        if (isVisible) View.VISIBLE else View.GONE)
  }

  fun setCustomerDebtAfterPayment(debt: BigDecimal?, @ColorRes textColor: Int, isVisible: Boolean) {
    _fragment.fragmentBinding.productOrder.customerDebt.setText(
        debt?.let {
          CurrencyFormat.format(it, AppCompatDelegate.getApplicationLocales().toLanguageTags())
        })
    _fragment.fragmentBinding.productOrder.customerDebt.setTextColor(
        _fragment.requireContext().getColor(textColor))
    _fragment.fragmentBinding.productOrder.customerDebt.setVisibility(
        if (isVisible) View.VISIBLE else View.GONE)
  }

  fun setInputtedProductOrders(productOrders: List<ProductOrderModel>) {
    val listLayout: LinearLayout =
        _fragment.fragmentBinding.productOrder.listLayout.apply {
          if (childCount > productOrders.size) {
            // Remove all the extra views.
            removeViews(productOrders.size, childCount - productOrders.size)
          }
        }
    for ((i, productOrder) in productOrders.withIndex()) {
      val cardBinding: ProductOrderCardBinding =
          if (listLayout.getChildAt(i) is MaterialCardView) {
                // Reuse already inflated view to reduce overhead.
                ProductOrderCardBinding.bind(listLayout.getChildAt(i))
              } else {
                ProductOrderCardBinding.inflate(_fragment.layoutInflater, listLayout, false)
              }
              .apply {
                checkbox.setChecked(false)
                checkbox.setVisibility(View.GONE)
                productImage.text.setVisibility(View.VISIBLE)
                cardView.setOnClickListener {
                  val cardIndex: Int = listLayout.indexOfChild(cardView)
                  if (_fragment.createQueueViewModel.selectProductOrderView.uiState.safeValue
                      .isContextualModeActive) {
                    _fragment.createQueueViewModel.selectProductOrderView
                        .onProductOrderCheckedChanged(cardIndex, !checkbox.isChecked)
                  } else {
                    _fragment.createQueueViewModel.makeProductOrderView.onShowDialog(
                        _fragment.createQueueViewModel.uiState.safeValue.productOrders[cardIndex])
                  }
                }
                cardView.setOnLongClickListener {
                  _fragment.createQueueViewModel.selectProductOrderView
                      .onProductOrderCheckedChanged(
                          listLayout.indexOfChild(cardView), !checkbox.isChecked)
                  true
                }
              }
      ProductOrderCardComponent(_fragment.requireContext(), cardBinding)
          .setProductOrder(productOrder)
      // Add to layout until the amount of child view match the product orders size.
      if (listLayout.childCount <= i) listLayout.addView(cardBinding.root)
    }
  }
}

private class SelectProductOrderActionModeCallback(private val _fragment: CreateQueueFragment) :
    ActionMode.Callback {

  override fun onCreateActionMode(mode: ActionMode?, menu: Menu?): Boolean {
    mode?.menuInflater?.inflate(R.menu.createqueue_contextualtoolbar, menu)
    return true
  }

  override fun onPrepareActionMode(mode: ActionMode?, menu: Menu?): Boolean = false

  override fun onActionItemClicked(mode: ActionMode?, item: MenuItem?): Boolean =
      when (item?.itemId) {
        R.id.createqueue_contextualtoolbar_delete_item -> {
          _fragment.createQueueViewModel.selectProductOrderView.onDeleteSelectedProductOrder()
          true
        }
        else -> false
      }

  override fun onDestroyActionMode(mode: ActionMode?) {
    _fragment.createQueueViewModel.selectProductOrderView.onDisableContextualMode()
  }
}
